src/gui/dialogs/campaign_selection.cpp

```diff
@@ -149,17 +150,46 @@ void campaign_selection::sort_campaigns(window& window, campaign_selection::CAMP
 	tree_view& tree = find_widget<tree_view>(&window, "campaign_tree", false);
 
 	// Remember which campaign was selected...
-	std::string was_selected = tree.selected_item()->id();
+	std::string was_selected;
+	if(!tree.empty()) {
+		was_selected = tree.selected_item()->id();
+		tree.clear();
+	}
+
+	boost::dynamic_bitset<> show_items;
+	show_items.resize(levels.size(), true);
+
+	if(!last_search_words_.empty()) {
+		for(unsigned i = 0; i < levels.size(); ++i) {
+			bool found = false;
+			for(const auto& word : last_search_words_) {
+				found = translation::ci_search(levels[i]->name(), word);
 
-	tree.clear();
+				if(!found) {
+					break;
+				}
+			}
 
-	for(const auto& level : levels) {
-		add_campaign_to_tree(window, level->data());
+			show_items[i] = found;
+		}
 	}
 
-	if(!was_selected.empty()) {
+	bool exists_in_filtered_result = false;
+	for(unsigned i = 0; i < levels.size(); ++i) {
+		if(show_items[i]) {
+			add_campaign_to_tree(window, levels[i]->data());
+
+			if (!exists_in_filtered_result) {
+				exists_in_filtered_result = levels[i]->id() == was_selected;
+			}
+		}
+	}
+
+	if(!was_selected.empty() && exists_in_filtered_result) {
 		find_widget<tree_view_node>(&window, was_selected, false).select_node();
 	}
+
+	campaign_selected(window);
```

Comment:
This function call need for display the details of the selected campaign. Otherwise, the details of the campaign that was selected before applying the filter will be displayed.

Reply:
Thanks, I see the problem this is solving. What I really meant by my question is why the code before this PR doesn't call campaign_selected(). I think the answer to that is because the select_node() call a few lines above would call campaign_selected() anyway because of these lines:

https://github.com/wesnoth/wesnoth/blob/31fa9fa33fc35673f03357b93ad0b8763739f6a9/src/gui/dialogs/campaign_selection.cpp#L202-L205

I suppose we should therefore call campaign_selected() only in an `else` branch of that `if`.

Reply:
Thank you for the detailed explanation. I fix it.